### PR TITLE
RANFS for contact

### PR DESCRIPTION
**Pull request by lindsayad**
Credit to @fdkong for the name: Reduced Active Nonlinear Function Set (RANFS). The algorithm is like so:

1. Start with the standard set of residual equations.
2. Check constraints
    - For an equality constraint we always apply
    - For an inequality constraint, we do a min comparison the inequality (gap for contact) and its associated Lagrange Multiplier.  If the inequality is less than the Lagrange Multiplier than we apply
3. Constraint application is done by replacing the non-linear residual equation with the constraint equation, e.g. for normal mechanical contact we replace the non-linear residual simply with the gap; in this way the gap is required to be driven to zero in an exact way. The replacement of the non-linear residual equation or non-linear function with the constraint equation reduces the total number of non-linear functions, hence the name of the scheme (RANFS)

Advantages of this scheme:
- No Lagrange Multipliers, e.g. no saddle-point and consequently we are free to use preconditioners like AMG
- No penalty term, so no introduction of ill conditioning into the matrix
- Exact satisfaction of the constraints

I've updated the frictionless contact algorithm table with the `RANFS` result (and also with timings from my current hardware). The `RANFS` scheme is the fastest both in terms of simulation time and non-linear iterations.

| Constraint | Displacement | NCP function | Time (arbitrary units) | Time steps | Nonlinear iterations |
| ---------- | ------------ | ------------ | ---------------------- | ---------- | -------------------- |
| Nodal | Mortar | Min | 4.164 | 40 | 104 |
| Nodal | Mortar | FB | 5.020 | 40 | 135 |
| Nodal | Nodal | Min | 3.124 | 41 | 104 |
| Nodal | Nodal | FB | 4.014 | 41 | 149 |
| Mortar | Mortar | Min | 4.461 | 40 | 106 |
| Mortar | Mortar | FB | 5.577 | 40 | 136 |
| Nodal | Nodal | RANFS | 2.700 | 40 | 99 |

### CODE REVIEW

**Comment by lindsayad** [comment]
The `RANFSNormalMechanicalContact` object includes ping-ponging protection. It detects when a slave node has alternated back and forth between two different element faces a sufficient number of times (currently face1-face2-face1-face2-face1) and then fixes the issue by applying more RANFS 😄 . Explicitly, I apply a number of equality constraints equal to the mesh dimension tying together the locations of the slave node and the nearest master node (which I assert is a node that both master faces involved in the ping-pong share). This ping-ponging protection is necessary for solving step 19 of the `bouncing-block-ranfs` test. Without the protection the solve does not converge and the time-step gets cut.

**Comment by lindsayad** [comment]
Magpie failure is due to move of `StandardType` into the `TIMPI` namespace. See libmesh/libmesh#2320. @dschwen I can update Magpie after this is merged (or if a standalone libmesh update is merged ahead of this)

File: modules/contact/doc/content/modules/contact/index.md
Comment (tophmatthews) [inline modules/contact/doc/content/modules/contact/index.md]: 👀

File: modules/contact/src/constraints/RANFSNormalMechanicalContact.C
Comment (fdkong) [inline modules/contact/src/constraints/RANFSNormalMechanicalContact.C]: No way to consider all normal components? Should we drive the gap along the normal direction to zero?

File: modules/contact/src/linesearches/PetscPingPongLineSearch.C
Comment (fdkong) [inline modules/contact/src/linesearches/PetscPingPongLineSearch.C]: How do you figure out it is the solution update? Should we change the function directly so that the update is right automatically?

File: modules/contact/test/src/constraints/RANFSTieNode.C
Comment (fdkong) [inline modules/contact/test/src/constraints/RANFSTieNode.C]: Potentially a dead lock .

File: modules/contact/test/tests/bouncing-block-contact/bouncing-block-ranfs.i
Comment (fdkong) [inline modules/contact/test/tests/bouncing-block-contact/bouncing-block-ranfs.i]: Cool! AMG works

File: modules/contact/test/tests/bouncing-block-contact/frictionless-nodal-reduced-active-set.i
Comment (fdkong) [inline modules/contact/test/tests/bouncing-block-contact/frictionless-nodal-reduced-active-set.i]: Not equal to 1.0?

File: modules/contact/test/tests/bouncing-block-contact/frictionless-nodal-reduced-active-set.i
Comment (fdkong) [inline modules/contact/test/tests/bouncing-block-contact/frictionless-nodal-reduced-active-set.i]: why use SVD? It is really hard to solve?

File: modules/contact/test/tests/bouncing-block-contact/tied-nodes.i
Comment (fdkong) [inline modules/contact/test/tests/bouncing-block-contact/tied-nodes.i]: SVD?

**Review by fdkong — CHANGES_REQUESTED** [review]

**Comment by fdkong** [comment]
It is a really cool algorithm. I am going to pull out the code and run it!

File: modules/contact/src/constraints/RANFSNormalMechanicalContact.C
Comment (lindsayad) [reply modules/contact/src/constraints/RANFSNormalMechanicalContact.C]: It doesn't have to be first order. Looping over nodes gives non-vertex nodes as well.

But yes it does have to be Lagrange

File: modules/contact/src/constraints/RANFSNormalMechanicalContact.C
Comment (lindsayad) [reply modules/contact/src/constraints/RANFSNormalMechanicalContact.C]: So should I just do `->close` since it needs to be closed? What's the right way to do this?

Comment (lindsayad) [reply modules/contact/src/constraints/RANFSNormalMechanicalContact.C]: I guess I could do an a `_comm.max` and then decide whether to close based on that communication...?

File: modules/contact/src/constraints/RANFSNormalMechanicalContact.C
Comment (lindsayad) [reply modules/contact/src/constraints/RANFSNormalMechanicalContact.C]: It's definitely a heuristic. In my testing experience, if a node ping-pongs back and forth 5 times then it will ping pong indefinitely. However, if it goes 3 times for example, it is not guaranteed to ping-pong indefinitely and the Newton iteration may naturally resolve the correct face dofs that need to be involved in the constraint.

File: modules/contact/src/constraints/RANFSNormalMechanicalContact.C
Comment (lindsayad) [reply modules/contact/src/constraints/RANFSNormalMechanicalContact.C]: You only want to introduce one constraint: the gap is zero. By definition, then we can only eliminate one non-linear function. We cannot do such a thing as eliminate n_x of one nonlinear function and n_y of another nonlinear function. We have to make a choice

File: modules/contact/test/tests/bouncing-block-contact/frictionless-nodal-reduced-active-set.i
Comment (lindsayad) [reply modules/contact/test/tests/bouncing-block-contact/frictionless-nodal-reduced-active-set.i]: This is just to verify that the setup is non-singular

Comment (lindsayad) [reply modules/contact/test/tests/bouncing-block-contact/frictionless-nodal-reduced-active-set.i]: I could perhaps change the default `-pc_type` here and then just specify `cli_args` from the `tests` spec

File: modules/contact/src/linesearches/PetscPingPongLineSearch.C
Comment (lindsayad) [reply modules/contact/src/linesearches/PetscPingPongLineSearch.C]: I might have to call and talk to you about this one. Too much to type

File: modules/contact/src/constraints/RANFSNormalMechanicalContact.C
Comment (fdkong) [reply modules/contact/src/constraints/RANFSNormalMechanicalContact.C]: Just delete `if (!_jacobian->closed())`. Do not need to any extra thing. PETSc will figure out

If you think you need to close, just close without any checks

File: modules/contact/src/constraints/RANFSNormalMechanicalContact.C
Comment (fdkong) [reply modules/contact/src/constraints/RANFSNormalMechanicalContact.C]: It will be helpful to put these comments into the code

File: modules/contact/src/constraints/RANFSNormalMechanicalContact.C
Comment (fdkong) [reply modules/contact/src/constraints/RANFSNormalMechanicalContact.C]: I see

File: modules/contact/src/constraints/RANFSNormalMechanicalContact.C
Comment (lindsayad) [reply modules/contact/src/constraints/RANFSNormalMechanicalContact.C]: I'm adding a check in the constructor that will error out if the displacement variables don't use a Lagrange basis

File: modules/contact/src/constraints/RANFSNormalMechanicalContact.C
Comment (lindsayad) [reply modules/contact/src/constraints/RANFSNormalMechanicalContact.C]: 👍

File: framework/src/problems/FEProblem.C
Comment (lindsayad) [reply framework/src/problems/FEProblem.C]: I added an error at the end of this method

File: framework/src/systems/NonlinearSystemBase.C
Comment (lindsayad) [reply framework/src/systems/NonlinearSystemBase.C]: K, changed the assertion to `slave_dofs.size() == slave_var.count()`

File: modules/contact/src/constraints/RANFSNormalMechanicalContact.C
Comment (lindsayad) [reply modules/contact/src/constraints/RANFSNormalMechanicalContact.C]: Done

File: modules/contact/test/tests/bouncing-block-contact/frictionless-nodal-reduced-active-set.i
Comment (lindsayad) [reply modules/contact/test/tests/bouncing-block-contact/frictionless-nodal-reduced-active-set.i]: Changed

File: modules/contact/test/tests/bouncing-block-contact/frictionless-nodal-reduced-active-set.i
Comment (lindsayad) [reply modules/contact/test/tests/bouncing-block-contact/frictionless-nodal-reduced-active-set.i]: K I changed the input file to use amg

File: modules/contact/test/tests/bouncing-block-contact/tied-nodes.i
Comment (lindsayad) [reply modules/contact/test/tests/bouncing-block-contact/tied-nodes.i]: Changed input file to use amg

**Comment by lindsayad** [comment]
@fdkong we will have to wait for the libmesh update before we can merge this

**Comment by lindsayad** [comment]
Ok libmesh update is in so if reviewers are ok with this, it can be merged (pending tests passing of course)

**Review by fdkong — DISMISSED** [review]

**Comment by lindsayad** [comment]
Let's not merge this until I determine what went wonky with the parallel testing

**Comment by fdkong** [comment]
> Let's not merge this until I determine what went wonky with the parallel testing

I almost clicked the button.

File: modules/contact/doc/content/modules/contact/index.md
Comment (tophmatthews) [inline modules/contact/doc/content/modules/contact/index.md]: `RANFS denotes the [Reduced Active Nonlinear Function Set](RANFSNormalMechanicalContact.md) scheme` ?

**Comment by lindsayad** [comment]
Ok @fdkong this is ready for the button!

**Review by fdkong — APPROVED** [review]

**Pull request merged**
